MB-8577 Devseed refactor to better support sub-scenarios

donaldthai opened the pull request:
## Description

Sub-scenarios are a recent addition to the `generate-test-data` script we use to seed our database. This PR refactors, cleans up and introduces a new pattern on how to seed the database in the `devseed.go` file. It should now be "easier" to add in a new sub-scenario, run a single sub-scenario and hopefully organizes the different seed data as we continue to add more cases. 

I figured I start with the `devseed.go` file since that's the one we're using to seed local data during development. The other scenario files `e2ebasic.go` and `bandwidth.go` are used for other cases.

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?
- I moved out the functions that helped seeded the database into the `shared.go` file since the functions were acting as helpers
- There is a new file `subscenarios.go` that stores all the subscenario functions that is added to the devseed list. Originally, I wanted to create a new sub-directory to keep things a bit more organize but ran into a circular reference when trying to clean up the different functions being used in the different named scenarios. So for now, we have the `subscenarios.go` file in the `scenario` directory.

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

Reset and migrate your local database
```sh
make db_dev_reset db_dev_migrate
```

Clear out binary and regenerate generate-test-data
```sh
rm -rf ./bin
make bin/generate-test-data
```

Run script. We'll use `shipment_hhg_cancelled` sub-scenario as an example. The full list can be found [here](https://github.com/transcom/mymove/blob/f8ad407828ac7755d79ff3556031dda172d6b355/pkg/testdatagen/scenario/devseed.go#L51).
```sh
generate-test-data --named-scenario="dev_seed" --named-sub-scenario="shipment_hhg_cancelled" --db-env="development"
```

You can check if the seeded data is present by running the app as a TOO:
```sh
make server_run
make office_client_run
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#logging)
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-8577) for this change
* [Wiki](https://github.com/transcom/mymove/wiki/Test-Data-Generation)

robot-mymove commented:
<!--
  0 failure: 
  0 warning: 
  1 messages
  
  DangerID: danger-id-Danger;
-->



<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Messages</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:book:</td>
      <td>:link: <a href="https://dp3.atlassian.net/browse/MB-8577">MB-8577</a></td>
    </tr>
  </tbody>
</table>


<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 36e030c2f364c7af5db1e8af9c6ccd245b36012f
</p>

donaldthai commented on cmd/generate-test-data/main.go:
```diff
@@ -254,7 +245,18 @@ func main() {
+			// Sub-scenarios are generated at run time
+			// Check config
+			// optional flag
+			if err = checkConfigNamedSubScenarioFlag(v, tdgs.NamedScenario(tdgs.DevSeedScenario)); err != nil {
+				logger.Fatal("invalid configuration", zap.Error(err))
+			}
```
This checks the `named-sub-scenario` flag after the data has been setup since the data is populated at run time.

donaldthai commented on pkg/testdatagen/scenario/devseed.go:
```diff
@@ -11,4173 +11,75 @@
-			},
-			UserUploader: userUploader,
-		})
+	// sets the sub-scenarios
+	e.SubScenarios = map[string]func(){
```
This is where the list is populated and setup

felipe-lee reviewed (approved):
These changes look good to me. I ran it and it works as stated. I really like having the functions that create the data being in a shared file. It makes the `devseed.go` file much easier to work with. Two comments:
* Can you update https://github.com/transcom/mymove/wiki/Test-Data-Generation#updating-scenarios with info on where things should be added/updated?
* It looks like you added error handling to some of the functions, which is surfacing some issues with `createPPMWithPaymentRequest` and `createPPMReadyToRequestPayment` (invalid status transitions). I'm not sure if that should be fixed before this gets merged or if those would be better to comment out in this pr and have a followup to get in there and properly fix them.

donaldthai commented:
> These changes look good to me. I ran it and it works as stated. I really like having the functions that create the data being in a shared file. It makes the `devseed.go` file much easier to work with. Two comments:
> 
> * Can you update https://github.com/transcom/mymove/wiki/Test-Data-Generation#updating-scenarios with info on where things should be added/updated?
> * It looks like you added error handling to some of the functions, which is surfacing some issues with `createPPMWithPaymentRequest` and `createPPMReadyToRequestPayment` (invalid status transitions). I'm not sure if that should be fixed before this gets merged or if those would be better to comment out in this pr and have a followup to get in there and properly fix them.

Just updated the wiki and fixed those two seed data functions! Thanks!

monfresh commented:
One issue I noticed is that some scenarios don't create the required users. For example, `subScenarioHHGOnboarding` doesn't create office users. Can we make sure all subscenarios create the appropriate users?

donaldthai commented:
> One issue I noticed is that some scenarios don't create the required users. For example, `subScenarioHHGOnboarding` doesn't create office users. Can we make sure all subscenarios create the appropriate users?

Okay, each sub-scenarios will have office users now. I also updated the Makefile for the `db_dev_e2e_populate` and `db_dev_fresh` to take in a variable `DEVSEED_SUBSCENARIO` to specify the sub-scenarios. We can run it like so: 
```bash
DEVSEED_SUBSCENARIO=shipment_hhg_cancelled make db_dev_e2e_populate
```

monfresh commented:
Sweet. Thanks! I tested it and it's working as expected.

monfresh reviewed (approved):
Great work! Thanks for making devseed easier to work with.
